org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/framework/protobuf/ProtobufSupport.java

```diff
@@ -0,0 +1,204 @@
+			JavaLanguageServerPlugin.getProjectsManager().getConnection().sendActionableNotification(notification);
+		}
+	}
+	
+	/**
+	 * Find all the Protobuf source output directories of the given project.
+	 * @param project project.
+	 */
+	private Set<File> findProtobufOutputDirectories(IProject project) {
+		IJavaProject javaProject = JavaCore.create(project);
+		Set<File> protobufOutputDirs = new HashSet<>();
+		try {
+			for (IClasspathEntry entry : javaProject.getRawClasspath()) {
+				if (entry.getEntryKind() != IClasspathEntry.CPE_SOURCE) {
+					continue;
+				}
+				for (IClasspathAttribute attribute : entry.getExtraAttributes()) {
+					if (attribute.getName().equals(PROTOBUF_GENERATED_SOURCE)
+							&& attribute.getValue().equals("true")) {
+						protobufOutputDirs.add(new File(project.getLocation().toFile(),
+								entry.getPath().removeFirstSegments(1).toString()));
+					}
+				}
+			}
+		} catch (JavaModelException e) {
+			JavaLanguageServerPlugin.log(e);
+		}
+		return protobufOutputDirs;
+	}
+
+	/**
+	 * Check if any of the input directories contains Java source files.
+	 * @param generatedDirectories directories to check.
+	 */
+	private boolean containsJavaFiles(Set<File> generatedDirectories) {
+		for (File dir : generatedDirectories) {
+			if (!dir.exists()) {
+				return false;
+			}
+
+			try (Stream<Path> walkStream = Files.walk(dir.toPath())) {
+				boolean containsJavaFile = walkStream.filter(p -> p.toFile().isFile()).anyMatch(f -> {
+					return f.toString().endsWith(".java");
+				});
+
+				if (containsJavaFile) {
+					return true;
+				}
+			} catch (IOException e) {
+				JavaLanguageServerPlugin.logException(e);
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Run generate proto tasks.
+	 * @param projectUris the project uris where the tasks will be executed.
+	 * @param monitor progress monitor.
+	 */
+	public static void generateProtobufSources(List<String> projectUris, IProgressMonitor monitor) {
+		JavaLanguageClient client = JavaLanguageServerPlugin.getProjectsManager().getConnection();
+		ProgressReport progressReport = new ProgressReport(UUID.randomUUID().toString());
+		progressReport.setTask("Running Gradle tasks");
+		progressReport.setComplete(false);
+		progressReport.setTotalWork(projectUris.size());
+		progressReport.setStatus("Generating Java sources from proto files...");
+		client.sendProgressReport(progressReport);
+		try {
+			boolean succeeded = false;
+			for (String projectUri : projectUris) {
+				if (!StringUtils.isEmpty(projectUri)) {
+					succeeded = runGenerateProtobufTasks(projectUri, monitor);
+				}
+				progressReport.setWorkDone(progressReport.getWorkDone() + 1);
+			}
+
+			if (!succeeded) {
+				ActionableNotification notification = new ActionableNotification().withSeverity(MessageType.Error)
+						.withMessage("Exception happens when generating source files, please open logs for details.")
+						.withCommands(Arrays.asList(new Command("Open", "java.open.serverLog")));
```

Comment:
"java.open.serverLog" is a vscode-java specific command, I would not to use it on server code directly. How about revealing specific exception details in the notification message instead of asking to open log file?

Reply:
appended a commit to make it fail saliently.